src/katgpucbf/xbgpu/engine.py

```diff
@@ -495,6 +854,39 @@ def populate_sensors(self, sensors: aiokatcp.SensorSet, rx_sensor_timeout: float
         self.add_service_task(time_sync_task)
         self._cancel_tasks.append(time_sync_task)
 
+    def free_rx_item(self, item: RxQueueItem) -> None:
+        """Return an RxQueueItem to the free queue if its refcount hits zero."""
+        item.refcount -= 1
+        if item.refcount == 0:
+            # All Pipelines are done with this item
+            self._active_in_sem.release()
+            # Need to wait for any outstanding events before recycling
```

Comment:
As usual, fair points all around. I don't like answering a question with another question, _but_ I think it may be an issue because
* Does the item still have an event to a/wait from `rx_item.add_marker(self._proc_command_queue)`?
* It's called/invoked just before the call to `free_rx_item` in the gpu_proc_loop.

Reply:
That captures the GPU processing done on the RxItem (the GPU memory), but you don't need to wait for that to reuse the chunk (CPU memory). Once you've copied the chunk to the GPU, the chunk is ready for reuse.